### PR TITLE
Feature/new sealing phase 2

### PR DESCRIPTION
**Pull request by durkmurder**

### CODE REVIEW

File: state/protocol/util/testing.go [hunk 1]
```diff
@@ -32,10 +44,25 @@ func RunWithFullProtocolState(t testing.TB, stateRoot *pbadger.StateRoot, f func
 		metrics := metrics.NewNoopCollector()
 		tracer := trace.NewNoopTracer()
 		consumer := events.NewNoop()
-		headers, _, seals, index, payloads, blocks, setups, commits, statuses := util.StorageLayer(t, db)
+		headers, _, seals, index, payloads, blocks, setups, commits, statuses, _ := util.StorageLayer(t, db)
+		state, err := pbadger.Bootstrap(metrics, db, headers, seals, blocks, setups, commits, statuses, stateRoot)
+		require.NoError(t, err)
+		receiptValidator := MockReceiptValidator()
```
Comment (durkmurder) [inline state/protocol/util/testing.go:50]: Comment for reviewers: I am using mocked `ReceiptValidator` in most tests since we are generating fixtures which are invalid if checked by `ReceiptValidator`. There is a separate test suite that tests the validator.

File: engine/consensus/matching/engine_test.go [hunk 2]
```diff
@@ -57,421 +52,187 @@ func TestMatchingEngine(t *testing.T) {
...
 		requestReceiptThreshold: 10,
 		maxResultsToRequest:     200,
-		assigner:                ms.assigner,
+		assigner:                ms.Assigner,
 		requireApprovals:        true,
+		receiptValidator:        ms.receiptValidator,
 	}
 }
 
-func (ms *MatchingSuite) TestOnReceiptInvalidOrigin() {
```
Comment (AlexHentschel) [inline engine/consensus/matching/engine_test.go:337]: checking Receipt validity is now handled by `ReceiptValidator`. For the matching engine, we only need to test that the output of `ReceiptValidator` is processed properly: see test https://github.com/onflow/flow-go/blob/32081ff9f91d24fd5c10ab57bdf447a70c7dd4ef/engine/consensus/matching/engine_test.go#L212-L213

File: engine/consensus/matching/engine_test.go [hunk 3]
```diff
@@ -57,421 +52,187 @@ func TestMatchingEngine(t *testing.T) {
...
 
-	ms.resultsDB.AssertNumberOfCalls(ms.T(), "Store", 0)
-	ms.resultsPL.AssertNumberOfCalls(ms.T(), "Add", 0)
-	ms.sealsPL.AssertNumberOfCalls(ms.T(), "Add", 0)
+	ms.ReceiptsPL.AssertNumberOfCalls(ms.T(), "Add", 0)
+	ms.ResultsPL.AssertNumberOfCalls(ms.T(), "Add", 0)
 }
 
-// try to submit a receipt for a known block from a consensus node
-func (ms *MatchingSuite) TestOnReceiptInvalidRole() {
```
Comment (AlexHentschel) [inline engine/consensus/matching/engine_test.go:369]: checking Receipt validity is now handled by `ReceiptValidator`. For the matching engine, we only need to test that the output of `ReceiptValidator` is processed properly: see test https://github.com/onflow/flow-go/blob/32081ff9f91d24fd5c10ab57bdf447a70c7dd4ef/engine/consensus/matching/engine_test.go#L212-L213

File: engine/consensus/matching/engine_test.go [hunk 4]
```diff
@@ -57,421 +52,187 @@ func TestMatchingEngine(t *testing.T) {
...
 
-	ms.resultsDB.AssertNumberOfCalls(ms.T(), "Store", 0)
-	ms.resultsPL.AssertNumberOfCalls(ms.T(), "Add", 0)
-	ms.sealsPL.AssertNumberOfCalls(ms.T(), "Add", 0)
+	ms.ResultsDB.AssertNumberOfCalls(ms.T(), "Store", 0)
+	ms.ResultsPL.AssertNumberOfCalls(ms.T(), "Add", 0)
 }
 
-// try ot submit a receipt from an Execution node, with zero unstaked node
-func (ms *MatchingSuite) TestOnReceiptUnstakedExecutor() {
```
Comment (AlexHentschel) [inline engine/consensus/matching/engine_test.go:386]: checking Receipt validity is now handled by `ReceiptValidator`. For the matching engine, we only need to test that the output of `ReceiptValidator` is processed properly: see test https://github.com/onflow/flow-go/blob/32081ff9f91d24fd5c10ab57bdf447a70c7dd4ef/engine/consensus/matching/engine_test.go#L212-L213

File: module/validation/receipt_validator.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+	return nil
+}
+
+// subgraphCheck enforces that result forms a valid sub-graph:
+// Let R1 be a result that references block A, and R2 be R1's parent result.
+// The execution results form a valid subgraph if and only if R2 references
+// A's parent.
+func (v *receiptValidator) subgraphCheck(result *flow.ExecutionResult) error {
+	prevResult, err := v.results.ByID(result.PreviousResultID)
```
Comment (zhangchiqing) [inline module/validation/receipt_validator.go:129]: What guarantees we have the results for `all` blocks in the storage?

The matching will store the result into the database when a receipt has been received and passed certain checks. However, this doesn't guarantee a consensus node have received the results for all the blocks, because each execution node only select a few consensus nodes to broadcast the receipts. 

It also doesn't seem to be the protocol state's `Extend` function is storing the result, did I miss anything?

File: engine/consensus/matching/engine.go [hunk 6]
```diff
@@ -262,30 +255,53 @@ func (e *Engine) onReceipt(originID flow.Identifier, receipt *flow.ExecutionRece
 		return nil
 	}
 
-	err = e.ensureStakedNodeWithRole(receipt.ExecutorID, head, flow.RoleExecution)
+	err = e.receiptValidator.Validate(receipt)
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply engine/consensus/matching/engine.go:258]: This is a good point Leo. Adding some thoughts here for further discussion:

* the `matching.Engine` re-requests receipts for finalized blocks that haven't been sealed yet (provided the backlog of unsealed finalized blocks is large enough) https://github.com/onflow/flow-go/blob/32081ff9f91d24fd5c10ab57bdf447a70c7dd4ef/engine/consensus/matching/engine.go#L891-L912
* So under the assumption that _all_ Execution nodes are honest and _always_ produce the correct result, I am thinking/hoping that it would be fine to drop Receipts for unknown blocks for the following reason:
  - having a single receipt for a block is sufficient for sealing liveness
  - if a consensus node does not have a receipt, it will eventually request it.  
* Caveat: Execution nodes do not always produce a correct result.
   - Therefore, we ideally need the ability to retrieve receipts from _all_ Execution Nodes. (as we might still miss the correct receipt). We can achieve this extending the code to request receipts for unsealed finalized blocks from all Execution Nodes from which we don't have a receipt yet.  Specifically adjusting this code: https://github.com/onflow/flow-go/blob/32081ff9f91d24fd5c10ab57bdf447a70c7dd4ef/engine/consensus/matching/engine.go#L882-L889

File: module/builder/consensus/builder.go [hunk 7]
```diff
@@ -197,167 +215,252 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
...
-			break
+
+		ancestorID = ancestor.ParentID
+	}
+
+	// Go through mempool and collect valid receipts. We store them by block
+	// height so as to sort them later. There can be multiple receipts per block
+	// even if they correspond to the same result.
+	receipts := make(map[uint64][]*flow.ExecutionReceipt) // [height] -> []receipt
+	for _, receipt := range b.recPool.All() {
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply module/builder/consensus/builder.go:420]: * we add receipts to this mempool in the `matching.Engine`: https://github.com/onflow/flow-go/blob/32081ff9f91d24fd5c10ab57bdf447a70c7dd4ef/engine/consensus/matching/engine.go#L263-L269 Receipts are added if they pass validation and are for unsealed blocks
* `matching.Engine` also cleans the mempools (see function `clearPools`). https://github.com/onflow/flow-go/blob/32081ff9f91d24fd5c10ab57bdf447a70c7dd4ef/engine/consensus/matching/engine.go#L813-L817 The specific criteria for removing a Receipt from the mempool are:
   *  once the block the receipt pertain to has a finalized seal: https://github.com/onflow/flow-go/blob/32081ff9f91d24fd5c10ab57bdf447a70c7dd4ef/engine/consensus/matching/engine.go#L800-L802 
   * or we are missing the respective blocks for 1000 iterations of the logic https://github.com/onflow/flow-go/blob/32081ff9f91d24fd5c10ab57bdf447a70c7dd4ef/engine/consensus/matching/engine.go#L788-L790 (very fuzzy logic, but I think sufficient for now)

File: module/validation/receipt_validator.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+	if result.Chunks.Len() != requiredChunks {
+		return engine.NewInvalidInputErrorf("invalid number of chunks, expected %d got %d",
+			requiredChunks, result.Chunks.Len())
+	}
+
+	return nil
+}
+
+// subgraphCheck enforces that result forms a valid sub-graph:
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply module/validation/receipt_validator.go:124]: This particular function does _not_ check whether the results form a chain. Instead, it verifies that the graph of results and receipts satisfies protocol requirements: 
Sealing_-_consensus_verification_responsibility_(2)
For details see [Protocol updates required to address Wintermute (notion)](https://www.notion.so/dapperlabs/Protocol-updates-required-to-address-Wintermute-0e6ea1f8c2dc4c9c81cd59a5fe3b4b8c#98f0cfad80214585b48cb960612ebeb2)

Nevertheless, your point is valid: where do we check that the results form a valid chain? 
* currently, we do this in the builder: https://github.com/onflow/flow-go/blob/32081ff9f91d24fd5c10ab57bdf447a70c7dd4ef/module/builder/consensus/builder.go#L339-L347
* Caution, this is a relaxed check: we only require that the end state matches the next results start state. However, we don't check that the `parent result ID` actually matches. 
  * We relaxed this check during our fire fighting, where execution nodes computed the ID of a result differently than consensus nodes. In that situation, the end state matched the next start state, but the IDs wouldn't match.
* Its a good point, we should check both conditions again. Thanks for flagging. 

#### ToDo:  

* _add_ the check to `receiptValidator` (maybe as a separate method?)
   -  the end state of the `PreviousResult` must match the start state of the current result   
   -  include unit test
* Replace the [check in the builder](https://github.com/onflow/flow-go/blob/32081ff9f91d24fd5c10ab57bdf447a70c7dd4ef/module/builder/consensus/builder.go#L339-L347):
   - for an `IncorporatedResultSeal` only check that `PreviousResultID` actually matches the result from the parent block's seal
* Urgency: exactly this logic runs currently on MainNet. We should fix it, but I don't think it needs to be addressed in this PR. Added to list of pending work: https://github.com/dapperlabs/flow-go/issues/5137

File: module/validation/receipt_validator.go (hunk 5 above)
Comment (AlexHentschel) [reply module/validation/receipt_validator.go:129]: Sorry, this is a bit of a subtle details. We should probably document it better:
* The `mutator`s store the block https://github.com/onflow/flow-go/blob/02803f99ffa185e22ddecaececa2075f8cf11345/state/protocol/badger/mutator.go#L650
* the storage implementation of `Blocks` stores the payload as a separate entity https://github.com/onflow/flow-go/blob/02803f99ffa185e22ddecaececa2075f8cf11345/storage/badger/blocks.go#L40
* where the payload is internally stored component by component (i.e. it also stores the receipts individually): https://github.com/onflow/flow-go/blob/02803f99ffa185e22ddecaececa2075f8cf11345/storage/badger/payloads.go#L52-L58
* and finally the receipts storage layer breaks up the receipt in the meta information and the result, where both are stored separately https://github.com/onflow/flow-go/blob/02803f99ffa185e22ddecaececa2075f8cf11345/storage/badger/receipts.go#L32

Furthermore, the matching engine also persists the `result` of any receipt that passes validation: https://github.com/onflow/flow-go/blob/32081ff9f91d24fd5c10ab57bdf447a70c7dd4ef/engine/consensus/matching/engine.go#L274-L276 

#### TLDR
* The storage layer has the `Result`s  incorporated in any known block or any received Execution Receipt (provided they pass validation)
* You are right that we discard Receipts for which we don't have the block or the parent result they reference. However, I don't think this is a problem for exactly the same reason explained in the [comment above](https://github.com/onflow/flow-go/pull/247#discussion_r545573100): we will eventually re-fetch them or other consensus nodes will include them in blocks. 🤞

File: engine/consensus/matching/engine.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (arrivets) [reply engine/consensus/matching/engine.go:258]: I think we can also rely on other consensus nodes to relay receipts. This is already the case because If a consensus node does not have a receipt, it is likely that another consensus node will have it and will include it in the chain, thereby indirecly sharing it with other consensus nodes.

**Review by arrivets — APPROVED** [review]

File: engine/consensus/matching/engine.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zhangchiqing) [reply engine/consensus/matching/engine.go:258]: OK, agreed to drop it and rely on the `requestPending`. And if so, we'd better add a comment to explain:
```suggestion
        // if the receipt is not valid, because the parent result is unknown, we will drop this receipt.
        // in the case where a child receipt is dropped because it is received before its parent receipt, we will
        // eventually request the parent receipt with the `requestPending` function
	err = e.receiptValidator.Validate(receipt)
```

File: module/validation/receipt_validator.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+// Let R1 be a result that references block A, and R2 be R1's parent result.
+// The execution results form a valid subgraph if and only if R2 references
+// A's parent.
+func (v *receiptValidator) subgraphCheck(result *flow.ExecutionResult) error {
+	prevResult, err := v.results.ByID(result.PreviousResultID)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return engine.NewInvalidInputErrorf("receipt's previous result (%x) is unknown", result.PreviousResultID)
```
Comment (zhangchiqing) [inline module/validation/receipt_validator.go:132]: When parent result is unknown, we aren't sure if it's really invalid. so a OutdatedInputError would be more accurate.
```suggestion
			return engine.NewOutdatedInputErrorf("receipt's previous result (%x) is unknown", result.PreviousResultID)
```

File: module/validation/receipt_validator.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (zhangchiqing) [reply module/validation/receipt_validator.go:124]: Thanks for reminding that. 

If we will eventually add all the checks to validate the results form a valid chain, we could still rename it as `resultChainCheck`, and adding TODO for missing checks to be added in the future.

`sub graph` feels like a new term to me that requires to understand its behavior.

File: module/validation/receipt_validator.go (hunk 5 above)
Comment (zhangchiqing) [reply module/validation/receipt_validator.go:129]: Let's say there are blocks `A <- B <- C (ER_A) <- D (ER_B) <- E`, `C` contains result for `A`.

So the happy case is: if we receive ER_C, whose parent result ID is the ID of  `ER_B`, then we will accept it

And if we receives `ER_D`, whose parent result ID is ID of `ER_C`, even if C is known, we will still drop this receipt, because we haven't receive ER_C yet. And we will rely on the `requestPending` to fetch it.

**Review by zhangchiqing — APPROVED** [review]

File: module/validation/receipt_validator.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply module/validation/receipt_validator.go:124]: 👍 
added this to the list of future revisions: https://github.com/dapperlabs/flow-go/issues/5137

File: module/validation/receipt_validator.go (hunk 9 above)
Comment (AlexHentschel) [reply module/validation/receipt_validator.go:132]: Agree. The error is not quite fitting. 

🤔 in my mind, "outdated" means too old (like the result is already sealed and finalized) ... but this could occur if the result is too new, se we haven't received the parent result yet. For blocks, we use `ErrUnverifiableBlock` in the event handler. Maybe we should include a similar error here ...

Comment (AlexHentschel) [reply module/validation/receipt_validator.go:132]: added ToDo item to https://github.com/dapperlabs/flow-go/issues/5153

Comment (zhangchiqing) [reply module/validation/receipt_validator.go:132]: yeah, agreed

**Review by AlexHentschel — APPROVED** [review]
I think this is finally done 😅
